evm_loader/lib/src/types/request_models.rs

```diff
@@ -102,6 +102,7 @@ pub struct EmulateRequestModel {
     #[serde(flatten)]
     pub emulation_params: EmulationParamsRequestModel,
     pub slot: Option<u64>,
+    pub write_version: Option<u64>,
```

Comment:
This parameter is actually only needed by `/api/trace` endpoint, so it should be moved to `TraceRequestModel`. I know it might look weird moving it there, but `emulate_request: EmulateRequestModel` is actually flattened (`#[serde(flatten)]`) into `TraceRequestModel`. In my opinion, another completely different struct should have been used instead of reusing the `EmulateRequestModel` fields in `TraceRequestModel`, to avoid situations like this one.

The `/api/emulate` endpoint is used to implement [`eth_call`](https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-eth#eth-call), which emulates a transaction on top of the block given by the `blockNrOrHash` parameter: https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L1165 . There is no need to specify a transaction index inside the block, since the blockchain state of the block itself is used as the execution base.

In comparison, tracing a transaction specified by hash needs to execute the transaction on the intermediate state of the block containing it, by recreating the state given by the transactions executed before it: https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/api.go#L827

So: 
- `/api/emulate` is used by `eth_call`: https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L1165
```go
state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
```
- `/api/trace` is used by `debug_traceCall` (tracing variant of `eth_call`): https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/api.go#L866
```go
statedb, release, err := api.backend.StateAtBlock(ctx, block, reexec, nil, true, false)
```
- `/api/trace` will also be used by `debug_traceTransaction`: https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/api.go#L827
```go
tx, blockHash, blockNumber, index, err := api.backend.GetTransaction(ctx, hash)
msg, vmctx, statedb, release, err := api.backend.StateAtTransaction(ctx, block, int(index), reexec)
```

Reply:
I agree with your reasoning, but Anton advised me to add this parameter to the `EmulateRequestModel` structure.

@anton-lisanin I would like to hear your opinion. Do you agree with @andreisilviudragnea?